tests/ui/deref_addrof.rs

```diff
@@ -72,3 +72,37 @@ impl S {
         //~^ deref_addrof
     }
 }
+
+fn issue14386() {
+    use std::mem::ManuallyDrop;
+
+    struct Data {
+        num: u64,
+    }
+
+    union DataWithPadding {
+        data: ManuallyDrop<Data>,
+        x: ManuallyDrop<u64>,
+        with_padding: [u8; size_of::<Data>()],
+        y: (u32, u32),
+        z: (ManuallyDrop<Data>,),
+    }
+
+    let mut a = DataWithPadding {
+        with_padding: [0; size_of::<DataWithPadding>()],
+    };
+    unsafe {
+        // Do not lint as `DerefMut` will not work on non-`Copy` union
+        // fields.
+        (*(&raw mut a.data)).num = 42;
+        (*(&raw mut a.z)).0.num = 42;
+
+        // Lint, as `Copy` fields in unions are ok.
+        (*(&raw mut a.with_padding)) = [1; size_of::<DataWithPadding>()];
+        //~^ deref_addrof
+        *(*(&raw mut a.x)) = 0;
+        //~^ deref_addrof
+        (*(&raw mut a.y)).0 = 0;
+        //~^ deref_addrof
```

Comment:
The error looks to be specific to `ManuallyDrop` rather than `Copy`, e.g. this is an error even though everything is `Copy`

```rust
#[derive(Copy, Clone)]
struct Data {
    num: u64,
}

union U {
    data: ManuallyDrop<Data>,
}

fn x(mut a: U) {
    unsafe {
        a.data.num = 3;
    }
}
```

`*a.x = 0;` doesn't hit it because it doesn't use `DerefMut`

Reply:
https://github.com/rust-lang/rust/blob/705421b52239d7393b4738764b192179d5c139c2/compiler/rustc_hir_typeck/src/place_op.rs#L311-L323